torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,485 @@
+        self._sync_param_groups()
+
+        local_cpu_state = _recursive_copy_to_device(
+            self.local_state_dict(), non_blocking=True, device=torch.device("cpu")
+        )
+
+        # Pull the sharded state from all the other replicas
+        # Store all the states in order, rank by rank
+        if self.rank == recipient_rank:
+            logging.debug("Pulling the sharded optimizer state from all replicas")
+            self._all_states = []
+            for rank in range(self.world_size):
+                if rank == self.rank:
+                    self._all_states.append(local_cpu_state)
+
+                    # Sync with other replicas, almost noop
+                    dist.broadcast_object_list([0], src=self.global_rank, group=self.group)
+                else:
+                    # Fetch the optim state from the other replicas
+                    global_rank = _get_global_rank(self.group, rank)
+                    replica_state = [0]
+                    dist.broadcast_object_list(replica_state, src=global_rank, group=self.group)
+
+                    self._all_states.append(
+                        _recursive_copy_to_device(replica_state[0], non_blocking=True, device=torch.device("cpu"))
+                    )
+        else:
+            # Acknowledge broadcasts, and send this rank's shard when needed
+            # Default to CPU space to gain some memory headroom
+            for rank in range(self.world_size):
+                if rank == self.rank:
+                    # Send the state to the reference replica
+                    dist.broadcast_object_list([local_cpu_state], src=self.global_rank, group=self.group)
+                else:
+                    global_rank = _get_global_rank(self.group, rank)
+
+                    # Discard this tensor/rank, broadcast necessary for syncing and because NCCL does not support gather
+                    dist.broadcast_object_list([0], src=global_rank, group=self.group)
+
+    def partition_parameters(self) -> List[List[Dict]]:
+        """Partitions parameters across distributed data parallel ranks.
+
+        Returns: a list of ``param_groups`` (which is a list of dict) where each
+        element of the list contains the param_groups for a rank. Element 0
+        corresponds to rank 0, etc. We need all the ranks for the broadcast
+        inside ``step()``.
+        """
+        if len(self._partition_parameters) == 0:
+            self._partition_parameters = [list() for _ in range(self.world_size)]
+            sizes = [0] * self.world_size
+            for param_group in self.param_groups:
+                param_lists: List[List] = [list() for _ in range(self.world_size)]
+                for param in param_group["params"]:
+                    # Add this param to rank with smallest size.
+                    rank = sizes.index(min(sizes))
+                    param_lists[rank].append(param)
+                    sizes[rank] += param.numel()
```

Comment:
this is neat :) 

Would I be correct if I assume even thorough this is not the most efficient algorithm, but it should be fine as in real training jobs params partitioning does not occur frequently?

Reply:
yes, that's the assumption at least ? Some frameworks are wiping things out often though (some GC thing), so it may well be a bottle neck in some cases. Ideally maybe that an exposed partitioning function would be ideal, but may be over complicating things

Reply:
> Ideally maybe that an exposed partitioning function would be ideal, but may be over complicating things

Agree, this is not urgent, we can add that after the release